go/ephemeral/ephemeral.go

```diff
@@ -188,13 +193,236 @@ func GetOwnDeviceEKs(ctx context.Context, g *libkb.GlobalContext) ([]keybase1.De
-		currentDeviceEKs = append(currentDeviceEKs, signedMetadata)
+		currentDeviceEKs[matchingDevice.DeviceID] = verifiedMetadata
 	}
 
 	return currentDeviceEKs, nil
 }
+
+// ===============
+// === UserEKs ===
+// ===============
+
+type UserEKSeed [EKSeedSize]byte
+
+func newUserEphemeralSeed() (seed UserEKSeed, err error) {
+	randomSeed, err := makeNewRandomSeed()
+	if err != nil {
+		return seed, err
+	}
+	return UserEKSeed(randomSeed), nil
+}
+
+func (s *UserEKSeed) DeriveDHKey() (key *libkb.NaclDHKeyPair, err error) {
+	derived, err := libkb.DeriveFromSecret(*s, libkb.DeriveReasonUserEKEncryption)
+	if err != nil {
+		return nil, err
+	}
+	keypair, err := libkb.MakeNaclDHKeyPairFromSecret(derived)
+	return &keypair, err
+}
+
+type UserEKBoxMetadata struct {
+	RecipientDeviceID   keybase1.DeviceID `json:"recipient_device_id"`
+	RecipientGeneration int               `json:"recipient_generation"`
+	Box                 string            `json:"box"`
+}
+
+func postNewUserEK(ctx context.Context, g *libkb.GlobalContext, sig string, boxes []UserEKBoxMetadata) error {
+	boxesJSON, err := json.Marshal(boxes)
+	if err != nil {
+		return err
+	}
+	apiArg := libkb.APIArg{
+		Endpoint:    "user/user_ek",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"sig":   libkb.S{Val: sig},
+			"boxes": libkb.S{Val: string(boxesJSON)},
+		},
+	}
+	_, err = g.GetAPI().Post(apiArg)
+	return err
+}
+
+func PublishNewUserEK(ctx context.Context, g *libkb.GlobalContext) (data keybase1.UserEkMetadata, err error) {
+	currentMerkleRoot, err := g.GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return data, err
+	}
+
+	// TODO: Read the actual generation from the userEK store.
+	generation := 1
+
+	seed, err := newUserEphemeralSeed()
+	if err != nil {
+		return data, err
+	}
+
+	// TODO: Store the key.
+
+	dhKeypair, err := seed.DeriveDHKey()
+	if err != nil {
+		return data, err
+	}
+	metadata := keybase1.UserEkMetadata{
+		Kid:        dhKeypair.GetKID(),
+		Generation: generation,
+		HashMeta:   currentMerkleRoot.HashMeta(),
+	}
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return data, err
+	}
+
+	// Sign the metadata blob with the latest PUK.
+	pukKeyring, err := g.GetPerUserKeyring()
+	if err != nil {
+		return data, err
+	}
+	signingKey, err := pukKeyring.GetLatestSigningKey(ctx)
+	if err != nil {
+		return data, err
+	}
+	signedPacket, _, err := signingKey.SignToString(metadataJSON)
+
+	// Box the seed up for each valid deviceEK.
+	ownDeviceEKs, err := GetOwnDeviceEKs(ctx, g)
+	if err != nil {
+		return data, err
+	}
+	boxes := []UserEKBoxMetadata{}
+	for deviceID, deviceEK := range ownDeviceEKs {
+		recipientKey, err := libkb.ImportKeypairFromKID(deviceEK.Kid)
+		if err != nil {
+			return data, err
+		}
+		// Encrypting with a nil sender means we'll generate a random sender private key.
+		box, err := recipientKey.EncryptToString(seed[:], nil)
```

Comment:
note to self: fix box endpoint to return sig to verify this seed matches the signed KID